src/Stratis.Bitcoin/P2P/PeerConnector.cs

```diff
@@ -218,33 +222,7 @@ private void Disconnect()
         /// Refer to <see cref="IPeerAddressManager.SelectPeerToConnectTo(PeerIntroductionType)"/> for details on how this is done.
         /// </para>
         /// </summary>
-        internal NetworkAddress FindPeerToConnectTo()
-        {
-            int groupFail = 0;
-
-            NetworkAddress peer = null;
-
-            while (groupFail < 50 && !this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
-            {
-                peer = this.peerAddressManager.SelectPeerToConnectTo(this.peerIntroductionType);
-                if (peer == null)
-                    break;
-
-                if (!peer.Endpoint.Address.IsValid())
-                    continue;
-
-                bool peerExistsInGroup = this.RelatedPeerConnector.GlobalConnectedNodes().Any(a => this.groupSelector(a).SequenceEqual(this.groupSelector(peer.Endpoint)));
-                if (peerExistsInGroup)
-                {
-                    groupFail++;
-                    continue;
-                }
-
-                break;
-            }
-
-            return peer;
-        }
+        internal abstract NetworkAddress FindPeerToConnectTo();
```

Comment:
Protected rather then Internal.
Internal prevents later implementers to reuse this technique

Reply:
Fixed

Reply:
Ah no @dangershony this actually needs to be on the IPeerConnector interface as public so that I can call it from the unit tests ;)

Reply:
ah right yes